Content.Server/Body/Systems/SkeletonBodyManagerSystem.cs

```diff
@@ -0,0 +1,187 @@
+                    {
+                        notFound = false;
+                        partList.Add(part.Owner);
+                    }
+                }
+                if (notFound)
+                {
+                    _popupSystem.PopupEntity(Loc.GetString("skeleton-reassemble-fail"), uid, Filter.Entities(uid));
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private void Reassemble(ReassembleDoAfterComplete args)
+        {
+            var uid = args.Uid;
+            var component = args.Component;
+
+            if (component.DNA == null)
+                return;
+
+            // Creates the new entity and transfers the mind component
+            var speciesProto = _prototype.Index<SpeciesPrototype>(component.DNA.Value.Profile.Species).Prototype;
+            var mob = _entities.SpawnEntity(speciesProto, _entities.GetComponent<TransformComponent>(component.Owner).MapPosition);
+
+            Get<SharedHumanoidAppearanceSystem>().UpdateFromProfile(mob, component.DNA.Value.Profile);
+            _entities.GetComponent<MetaDataComponent>(mob).EntityName = component.DNA.Value.Profile.Name;
+
+            if (TryComp<MindComponent>(uid, out var mindcomp) && mindcomp.Mind != null)
+                mindcomp.Mind.TransferTo(mob);
+
+            // Cleans up all the body part pieces
+            foreach (var entity in args.PartList)
+            {
+                _entities.DeleteEntity(entity);
+            }
+
+            _popupSystem.PopupEntity(Loc.GetString("skeleton-reassemble-success", ("user", mob)), mob, Filter.Entities(mob));
+        }
+
+        /// <summary>
+        /// Called before the skeleton entity is gibbed in order to save
+        /// the dna for reassembly later
+        /// </summary>
+        /// param name="uid"></param> the entity the mind is going to be transfered which also stores the DNA
+        /// <param name="body"></param> the entity whose DNA is being saved
+        public void UpdateDNAEntry(EntityUid uid, EntityUid body)
+        {
+            if (!TryComp<SkeletonBodyManagerComponent>(uid, out var skelBodyComp) ||
+                !TryComp<MindComponent>(body, out var mindcomp))
+                return;
+
+            if (mindcomp.Mind == null)
+                return;
+
+            if (mindcomp.Mind.UserId == null)
+                return;
+
+            var profile = (HumanoidCharacterProfile) _prefsManager.GetPreferences(mindcomp.Mind.UserId.Value).SelectedCharacter;
+            skelBodyComp.DNA = new ClonerDNAEntry(mindcomp.Mind, profile);
+        }
+
+        private sealed class ReassembleDoAfterComplete : EntityEventArgs
+        {
+            public readonly EntityUid Uid; //the entity being reassembled
+            public readonly EntityUid User; //the user performing the reassembly
+            public readonly SkeletonBodyManagerComponent Component;
```

Comment:
dont store components on doafter events, just get it from the entityuid in the handler.